File: src/WinLock.Android/app/src/main/java/com/winlock/parent/protocol/PairingQrPayload.kt
package com.winlock.parent.protocol

import java.net.URLDecoder
import java.util.Base64

/**
 * What the PC encodes into the pairing QR code. Mirrors
 * WinLock.Core.Pairing.PairingQrPayload.ToQrText/TryParse exactly:
 * `winlock-pair:v1:{deviceId:N}:{urlEscapedDisplayName}:{token}:{secretBase64}:{fingerprintHex}:{hostAndPort}`
 *
 * The scanned QR is itself the secure channel — nobody but someone looking at this screen
 * can read it — so the shared secret travels inside it directly and is never sent over the
 * network, before or after pairing.
 */
data class PairingQrPayload(
    val deviceId: String,
    val deviceDisplayName: String,
    val token: String,
    val secret: ByteArray,
    val certificateFingerprintHex: String,
    val hostAndPort: String,
) {
    companion object {
        fun tryParse(text: String): PairingQrPayload? {
            // hostAndPort is last and may itself contain a colon, so cap the split rather
            // than splitting on every colon in the string.
            val parts = text.split(":", limit = 8)
            if (parts.size != 8 || parts[0] != "winlock-pair" || parts[1] != "v1") return null

            val deviceId = GuidUtil.fromNFormat(parts[2]) ?: return null

            val secret = try {
                Base64.getDecoder().decode(parts[5])
            } catch (e: IllegalArgumentException) {
                return null
            }

            val displayName = try {
                URLDecoder.decode(parts[3], "UTF-8")
            } catch (e: Exception) {
                return null
            }

            return PairingQrPayload(
                deviceId = deviceId,
                deviceDisplayName = displayName,
                token = parts[4],
                secret = secret,
                certificateFingerprintHex = parts[6],
                hostAndPort = parts[7],
            )
        }
    }
}

File: src/WinLock.Core/Pairing/PairingQrPayload.cs
namespace WinLock.Core.Pairing;

/// <summary>
/// What the PC encodes into the pairing QR code. The scanned QR is itself the secure
/// channel — nobody but someone looking at this screen can read it — so the shared secret
/// travels inside it directly and is never sent over the network, before or after pairing.
/// </summary>
public sealed record PairingQrPayload(
    Guid DeviceId,
    string DeviceDisplayName,
    string Token,
    byte[] Secret,
    string CertificateFingerprintHex,
    string HostAndPort)
{
    public string ToQrText() =>
        $"winlock-pair:v1:{DeviceId:N}:{Uri.EscapeDataString(DeviceDisplayName)}:{Token}:" +
        $"{Convert.ToBase64String(Secret)}:{CertificateFingerprintHex}:{HostAndPort}";

    /// <summary>Inverse of <see cref="ToQrText"/> — shared by every controller-side
    /// implementation (the test stub today, an Android client eventually) so there's exactly
    /// one place that understands the QR wire format.</summary>
    public static bool TryParse(string qrText, out PairingQrPayload? payload)
    {
        payload = null;

        // HostAndPort is last and may itself contain a colon, so cap the split at the
        // number of fields rather than splitting on every colon in the string.
        var parts = qrText.Split(':', 8);
        if (parts.Length != 8 || parts[0] != "winlock-pair" || parts[1] != "v1")
            return false;

        if (!Guid.TryParseExact(parts[2], "N", out var deviceId))
            return false;

        byte[] secret;
        try
        {
            secret = Convert.FromBase64String(parts[5]);
        }
        catch (FormatException)
        {
            return false;
        }

        payload = new PairingQrPayload(
            deviceId,
            Uri.UnescapeDataString(parts[3]),
            parts[4],
            secret,
            parts[6],
            parts[7]);
        return true;
    }
}
